### PR TITLE
187186207/iter1 golden repo fixes

### PR DESCRIPTION
**Pull request by ArushC**
### [Pivotal Tracker Link][tracker]

[tracker]: https://www.pivotaltracker.com/story/show/your-story-id

## What this PR does:

Implements minor fixes that were suggested on [https://github.com/beautyjoy/BJC-Teacher-Tracker/pull/291](https://github.com/beautyjoy/BJC-Teacher-Tracker/pull/291)
 

This pull request fixes|implements (pick one...) ______.

#### Who authored this PR?

Arush, Perry, Jackson, Raz, Michael

### How should this PR be tested?

* Is there a deploy we can view?

* What do the specs/features test?
* Are there edge cases to watch out for?

#### Are there any complications to deploying this?

*Run "rails db:migrate" since there are schema changes (a new TO field)

*(Optional) Run rails db:reseed if you want to directly import the [updated email templates we made](https://github.com/beautyjoy/BJC-Teacher-Tracker/pull/292/files#diff-7e0436c5774a8f6f798727122bfc14661031feaed667529064143541b17d2a1d) (updated email content, new template, and the pre-filled to fields)

*(Alternative way to step 2) Manually update the existing email template with the new content.
After running step 1 (migrate), for the new to field, fill in {{teacher_email}}, {{teacher_personal_email}} as the template value, or some default emails, e.g. lmock@berkeley.edu, contact@bjc.berkeley.edu
Also need to update the template content for Denial Email and Request More Info Email, as they no longer use the original reason field, but the denial_reason or request_reason instead. Sample format can be found at the seed_data.rb we provided, such as:
  @deny_email = <<-DENY_EMAIL1
    <p>Dear [Recipient's Name],</p>
    <p>Thank you for your email. We have received your message and we are sorry to inform you that your application has been rejected.</p>
    <p>Best regards,</p>
    <p>[Your Name]</p>
    <p>Below, you can find the reason as to why it was rejected </p>
    <p>
      {{ denial_reason | strip_tags}}
    </p>
  DENY_EMAIL1

### Checklist:

- [x] Has this been deployed to a staging environment or reviewed by a customer?
- [x] Tag someone for code review (either a coach / team member)
- [x] I have renamed the branch to match PivotTracker's suggested one (necessary for BlueJay) (e.g. `michael/12345-add-new-feature` Any branch name will do as long as the story ID is there. You can use `git checkout -b [new-branch-name]`)

### CODE REVIEW

**Comment by perryzjc** [comment]
Looks good to me as well. It seems to fix everything mentioned in Professor Ball's Initial comments.

If anything else is needed, I think we need some notes to document what admins need to do after the PR, especially about the migration.

Based on the discussions we had in today's lecture with Michael (Friday, Mar 8), Michael suggested we just document the manual steps to update the email template (since there are only quite a few templates right now), such as the values to fill out for the new `to` field.

We could possibly write something for the `Are there any complications to deploying this?` section as below:

1. Run "rails db:migrate" since there are schema changes  (a new `TO` field)
2. (Optional) Run `rails db:reseed` if you want to directly import the [updated email templates we made](https://github.com/beautyjoy/BJC-Teacher-Tracker/pull/292/files#diff-7e0436c5774a8f6f798727122bfc14661031feaed667529064143541b17d2a1d) (updated email content, new template, and the pre-filled `to` fields)
3. (Alternative way to step 2) Manually update the existing email template with the new content.
    - After running step 1 (migrate), for the new `to` field, fill in `{{teacher_email}}, {{teacher_personal_email}}` as the template value, or some default emails, e.g. `lmock@berkeley.edu, contact@bjc.berkeley.edu`
    - Also need to update the template content for `Denial Email` and `Request More Info Email`, as they no longer use the original `reason` field, but the `denial_reason` or `request_reason` instead. Sample format can be found at the `seed_data.rb` we provided, such as:
```
  @deny_email = <<-DENY_EMAIL1
    <p>Dear [Recipient's Name],</p>
    <p>Thank you for your email. We have received your message and we are sorry to inform you that your application has been rejected.</p>
    <p>Best regards,</p>
    <p>[Your Name]</p>
    <p>Below, you can find the reason as to why it was rejected </p>
    <p>
      {{ denial_reason | strip_tags}}
    </p>
  DENY_EMAIL1
```

File: app/mailers/teacher_mailer.rb [hunk 1]
```diff
@@ -11,34 +11,38 @@ class TeacherMailer < ApplicationMailer
   def welcome_email(teacher)
     @teacher = teacher
     set_body
-    mail to: teacher.email_name,
+    set_recipients
```
Comment (cycomachead) [inline app/mailers/teacher_mailer.rb:14]: I am going to merge this, but `set_recipients` probably is best as a `before_action` since it's applied to all email templates.

File: app/views/main/_deny_modal.html.erb [hunk 2]
```diff
@@ -33,31 +26,38 @@
...
+                newAction = `/teachers/${teacherId}/deny`;
+                newTitle = 'Deny ' + teacherName;
+                inputName = 'denial_reason';
+                break;
+            case "request_info":
+                newAction = `/teachers/${teacherId}/request_info`;
+                newTitle = 'Request Info from ' + teacherName;
+                inputName = 'request_reason';
+                break;
+            default:
```
Comment (cycomachead) [inline app/views/main/_deny_modal.html.erb:48]:
```suggestion
            case "deny":
            default:
```

We can combine the two into one block.

File: features/step_definitions/web_steps.rb [hunk 3]
```diff
@@ -60,6 +60,18 @@ def with_scope(locator, &block)
   page.execute_script('$(tinymce.editors[0].setContent("' + value + '"))')
 end
 
+When(/^(?:|I )follow the first "([^"]*)" link$/) do |link_text|
```
Comment (cycomachead) [inline features/step_definitions/web_steps.rb:63]: In the future, I would probably not add to web_steps, which is kind of generic and not very app-specific.

File: spec/controllers/teachers_controller_spec.rb [hunk 4]
```diff
@@ -183,4 +183,24 @@
       expect(Teacher.find_by(email: "valid_example@valid_example.edu").not_reviewed?).to be true
     end
   end
+
+  context "flash a message after resend_welcome_email" do
+    it "succeeds when teacher is validated, sets success" do
+      ApplicationController.any_instance.stub(:current_user).and_return(Teacher.find_by(first_name: "Validated"))
+      validated_teacher = Teacher.find_by(first_name: "Validated")
```
Comment (cycomachead) [inline spec/controllers/teachers_controller_spec.rb:190]:
```suggestion
      validated_teacher = Teacher.find_by(first_name: "Validated")
      ApplicationController.any_instance.stub(:current_user).and_return(validated_teacher)
```

Arguably, this spec should use a factory (or fixture at the very least...) rather than a `find_by` call, but we should perform the call first.

File: spec/controllers/teachers_controller_spec.rb [hunk 5]
```diff
@@ -183,4 +183,24 @@
...
+      request.env["HTTP_REFERER"] = edit_teacher_path(validated_teacher.id)
+      post :resend_welcome_email, params: { id: validated_teacher.id }
+      expect(flash[:success]).to eq("Welcome email resent successfully!")
+      expect(response).to redirect_to(edit_teacher_path(validated_teacher.id))
+    end
+
+    it "fails when teacher is not validated, sets alert" do
+      ApplicationController.any_instance.stub(:current_user).and_return(Teacher.find_by(first_name: "Short"))
+      nonvalidated_teacher = Teacher.find_by(first_name: "Short")
```
Comment (cycomachead) [inline spec/controllers/teachers_controller_spec.rb:199]:
```suggestion
      nonvalidated_teacher = Teacher.find_by(first_name: "Short")
      ApplicationController.any_instance.stub(:current_user).and_return(nonvalidated_teacher)
```

(Same as above)

File: spec/models/email_templates_reseed_spec.rb [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+# This test suite verifies the reseeding behavior of EmailTemplate models to ensure
+# that email templates are correctly handled (i.e., updated or not-removal) without duplication
+# upon reseeding the database with new seed data. It tests the initial seeding, updates to
+# templates, and the non-removal of templates that are no longer present in the seed data.
+RSpec.describe "EmailTemplate Reseed Behavior", type: :model do
+  RSpec.shared_context "setup initial email templates" do
+    before do
+      # Clear the EmailTemplate table to ensure a clean slate for each test.
+      # This is safe to do because it's under test environment, not production
+      EmailTemplate.delete_all
```
Comment (cycomachead) [inline spec/models/email_templates_reseed_spec.rb:14]: I am going to merge this PR, but this is an anti-pattern. We shouldn't specifically need to clear+re-seed the table. We should be using fixtures, or loading specific email templates.

File: spec/mailers/teacher_mailer_spec.rb [hunk 7]
```diff
@@ -36,4 +47,17 @@
...
   end
+
+  it "Sends Request Info Email" do
+    teacher = teachers(:long)
+    email = TeacherMailer.request_info_email(teacher, "Request Reason")
+    email.deliver_now
+    expect(email.from[0]).to eq("contact@bjc.berkeley.edu")
+    expect(email.to[0]).to eq("short@long.com")
+    expect(email.subject).to eq("Request Info Email")
+    # Test appearance of first_name
```
Comment (cycomachead) [inline spec/mailers/teacher_mailer_spec.rb:58]: This is a fair comment, but it'd be better to do something like 
expect body to **not** include `"first_name"`, or to be safe: expect that a template includes that string, then assert it is missing from the rendered email.

**Review by cycomachead — COMMENTED** [review]
Thank you! This looks great.

Almost all of my comments are pretty minor. I'm going to use GitHub's suggestions to and if CI is good, merge then deploy. You will want to make sure to pull from upstream/main after this is merged.
